### ISSUE TITLE
Inferencing Speeds

### ISSUE THREAD

**Issue by rhalder2023** [ultralytics/ultralytics#17242]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello,

So I am trying to use yolov8 on a real time project, that uses realsense camera feeds to perform predicitons on. The model im using is the yolov8s-seg.

I have a 12th gen i7 CPU, 20 cores; A2000 gpu with a 12gb virtual memory and have 24 gigs of RAM. 

When my model is running my cpu utilization is at  80%, and gpu utilization is around 90% as I am doing a lot of torch operations and image processing on the gpu. 

The model predictions are taking a lot of time, i am running 2 instances of the model each instance getting 2 images to predict, on a 60 fps loop. The model takes 0.03 to 0.06 ms to make a prediction, which is reducing my overall loop rate. I was kinda thinking my cpu is throttling the model, because even when i take away my torch functions, with high cpu usage the model takes longer to predict. 

Does anyone have any advice as to why this might be happening, and any suggestions on how to make it better?

I am running it on python3.10. 

If switching to tensorrt with c++ makes the inferencing faster, i am also down for that. Also i am running half precision, but that doesnt change the speed by much to see any noticeable difference in the overall loop rate of my prediction node.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @rhalder2023, thank you for reaching out to the Ultralytics community 🚀! Your setup for using `yolov8` with real-time predictions sounds fascinating. We appreciate you detailing the current performance metrics and hardware specs.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us debug it, as having a clear set of steps will allow us to understand the situation better.

For performance-related ❓ Questions, consider optimizing the workload on your CPU and GPU. Techniques such as batch processing or multithreading can be helpful. Feel free to share logs or further configuration you might have set; it helps us see the full picture. Exploring alternatives like TensorRT with C++ as you mentioned could potentially improve inference speeds, and sharing your experiences in this transition can benefit the community.

We recommend a quick visit to the [Docs](https://docs.ultralytics.com) for more specific optimization strategies and our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to possibly enhance your model's efficiency.

For further community support, you can join us on [Discord](https://ultralytics.com/discord) 🎧 for real-time advice, or explore threads on [Discourse](https://community.ultralytics.com) and [Subreddit](https://reddit.com/r/ultralytics) where many enthusiasts like you discuss various issues and solutions.

## Upgrade

Ensure you are using the latest version of the `ultralytics` package along with the necessary [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml). Upgrade your environment to enjoy recent performance improvements:

```bash
pip install -U ultralytics
```

## Environments

If you're open to new environments, try our up-to-date verified environments:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

Check the CI status, and ensure all systems are go: <a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

This is an automated response, and an Ultralytics engineer will attend to your query soon, providing further assistance.

**Comment by Y-T-G** [comment]
Are you currently using TensorRT?

You should also try and move the stream decoding to GPU using NVDEC.

**Comment by Y-T-G** [comment]
Also do you really need 60FPS? You can get away with 20FPS or less in a lot of "real-time" projects. Dropping those extra frames would free up a lot of compute.

**Comment by Y-T-G** [comment]
> 0.03 to 0.06 ms

Is that ms or s (30 to 60ms)?

**Comment by rhalder2023** [comment]
> Are you currently using TensorRT?
> 
> You should also try and move the stream decoding to GPU using NVDEC.

No, i am not, like not even the python tensorrt api, i am using the ultralytics predict function

**Comment by rhalder2023** [comment]
> Also do you really need 60FPS? You can get away with 20FPS or less in a lot of "real-time" projects. Dropping those extra frames would free up a lot of compute.

Yeah youre not wrong, so i had to launch realsense at 30 fps so that my model also runs at 30 fps, if realsense runs at 60 my model fps drops to 12 or 8, idk the computer is trying to compensate i think

**Comment by rhalder2023** [comment]
> > 0.03 to 0.06 ms
> 
> Is that ms or s (30 to 60ms)?

millisecond, like 0.03 ms is 30 fps

**Comment by Y-T-G** [comment]
> millisecond, like 0.03 ms is 30 fps

33 ms is 30 FPS. 0.03 ms would be 1000/0.03 = 33333 FPS

**Comment by Y-T-G** [comment]
> Yeah, so if i just use the .engine model, will it make it infernece faster, i thought i needed to explicitly use the tensorrt api that nvidia provides if i want to make full use of the nvidia gpus

Yes. `ultralytics` uses TensorRT API underneath.

**Comment by rhalder2023** [comment]
> > Yeah, so if i just use the .engine model, will it make it infernece faster, i thought i needed to explicitly use the tensorrt api that nvidia provides if i want to make full use of the nvidia gpus
> 
> Yes. `ultralytics` uses TensorRT API underneath.

ooohhhhhhh, I shall try it. Thanks You

**Comment by rhalder2023** [comment]
> > Yeah, so if i just use the .engine model, will it make it infernece faster, i thought i needed to explicitly use the tensorrt api that nvidia provides if i want to make full use of the nvidia gpus
> 
> Yes. `ultralytics` uses TensorRT API underneath.

umm when im trying to export pt to engine it says

"if fmt == "engine" and "dla" in self.args.device:
TypeError: argument of type 'NoneType' is not iterable

**Comment by rhalder2023** [comment]
> > > Yeah, so if i just use the .engine model, will it make it infernece faster, i thought i needed to explicitly use the tensorrt api that nvidia provides if i want to make full use of the nvidia gpus
> > 
> > 
> > Yes. `ultralytics` uses TensorRT API underneath.
> 
> umm when im trying to export pt to engine it says
> 
> "if fmt == "engine" and "dla" in self.args.device: TypeError: argument of type 'NoneType' is not iterable

Do i need to have tensorrt installed in my system? :3

**Comment by Y-T-G** [comment]
It should install automatically

**Comment by rhalder2023** [comment]
> It should install automatically

Yeah it is doing it in another system i have, that one it showed that error

**Comment by Y-T-G** [comment]
You can try installing it manually.
